src/NifFile.cpp

```diff
@@ -949,455 +949,451 @@ void NifFile::Optimize() {
 OptResult NifFile::OptimizeFor(OptOptions& options) {
 	OptResult result;
 
-	if (options.targetVersion.IsSSE() && hdr.GetVersion().IsSK()) {
-		// SK -> SSE
-		if (!isTerrain)
-			result.dupesRenamed = RenameDuplicateShapes();
+	const bool toSSE = options.targetVersion.IsSSE() && hdr.GetVersion().IsSK();
+	const bool toLE = options.targetVersion.IsSK() && hdr.GetVersion().IsSSE();
 
-		hdr.SetVersion(options.targetVersion);
+	if (!toSSE && !toLE) {
+		result.versionMismatch = true;
+		return result;
+	}
+
+	if (!isTerrain)
+		result.dupesRenamed = RenameDuplicateShapes();
 
-		for (auto& shape : GetShapes()) {
+	hdr.SetVersion(options.targetVersion);
+
+	auto shapes = GetShapes();
+	if (toSSE) {
+		std::for_each(shapes.begin(), shapes.end(), [this, &options, &result](auto* shape) {
 			std::string shapeName = shape->name.get();
 
 			auto geomData = hdr.GetBlock<NiGeometryData>(shape->DataRef());
-			if (geomData) {
-				bool removeVertexColors = true;
-				bool hasTangents = geomData->HasTangents();
-				std::vector<Vector3>* vertices = &geomData->vertices;
-				std::vector<Vector3>* normals = &geomData->normals;
-				const std::vector<Color4>& colors = geomData->vertexColors;
-				std::vector<Vector2>* uvs = nullptr;
-				if (!geomData->uvSets.empty())
-					uvs = &geomData->uvSets[0];
-
-				std::vector<Triangle> triangles;
-				geomData->GetTriangles(triangles);
-
-				if (!options.removeParallax)
-					removeVertexColors = false;
-
-				// Only remove vertex colors if all are 0xFFFFFFFF
-				if (removeVertexColors) {
-					Color4 white(1.0f, 1.0f, 1.0f, 1.0f);
-					for (auto& c : colors) {
-						if (white != c) {
-							removeVertexColors = false;
-							break;
-						}
+
+			if (!geomData)
+				return;
```

Comment:
Shouldn't this (and the SE counterpart) be `continue` instead of `return`?

Reply:
Yes, with a normal loop. `std::for_each` calls the lambda for each element, so in order to skip an element, you have to return out of the lambda